src/ariel-os-debug/src/lib.rs

```diff
@@ -125,69 +130,58 @@ mod backend {
 
 pub use backend::*;
 
-#[cfg(feature = "defmt")]
-pub mod log {
-    pub use defmt;
-
-    #[macro_export]
-    macro_rules! __trace {
-        ($($arg:tt)*) => {{
-            use $crate::log::defmt;
-            defmt::trace!($($arg)*);
-        }};
-    }
-
-    #[macro_export]
-    macro_rules! __debug {
-        ($($arg:tt)*) => {{
-            use $crate::log::defmt;
-            defmt::debug!($($arg)*);
-        }};
-    }
+#[doc(hidden)]
+#[cfg(feature = "log")]
+mod logger {
+    use log::{Level, LevelFilter, Metadata, Record};
+
+    static LOGGER: DebugLogger = DebugLogger;
+
+    const MAX_LEVEL: LevelFilter = {
+        let max_level =
+            ariel_os_utils::str_from_env_or!("DEBUG_LOG_LEVEL", "info", "maximum level to log");
+
+        // NOTE: these magic strings could likely be replaced with calls to
+        // `LevelFilter::*::as_str()` if that method was const.
+        if const_str::compare!(==, max_level, "trace") {
+            LevelFilter::Trace
+        } else if const_str::compare!(==, max_level, "debug") {
+            LevelFilter::Debug
+        } else if const_str::compare!(==, max_level, "info") {
+            LevelFilter::Info
+        } else if const_str::compare!(==, max_level, "warn") {
+            LevelFilter::Warn
+        } else if const_str::compare!(==, max_level, "error") {
+            LevelFilter::Error
+        } else if const_str::compare!(==, max_level, "off") {
+            LevelFilter::Off
+        } else if const_str::compare!(==, max_level, "") {
+            // Default level
+            LevelFilter::Info
+        } else {
+            panic!("invalid log level");
```

Comment:
can this be a compile-time error? should be all `const` ...
If not, can we, instead of panicking, default to `info` and `warn!("invalid log level {max_level}, defaulting to \"info\"?

Reply:
perfect!